### DISCUSSION TITLE
How to get the Last Inserted Id from this lib?

### DISCUSSION THREAD

**Discussion by ranganathanrajeshkumar** [vrogier/ocilib#357] [Q&A] [answered]
Hi,

OCCI lib supports getting the last inserted ID as below
stmt->registerOutParam(1, oracle::occi::Type::OCCINUMBER);

But The Statment::Register() method in this library has not option to get the same. Can you please guide how to achieve this?

**Comment by vrogier** [comment, marked as answer]
Hi,

OCCI does not provide automatic retrieval of the last inserted ID.

For that you must use a returning into clause that is also supported by OCILIB.

See OCILIB documentation:
https://vrogier.github.io/ocilib/doc/html/group___ocilib_c_api_feature_returning_into.html

C Demo sample :
https://github.com/vrogier/ocilib/blob/master/demo/returning.c

C++ Demo sample :
https://github.com/vrogier/ocilib/blob/master/demo/returning.cpp


Regards,

Vincent

**Comment by ranganathanrajeshkumar** [comment]
Hi @vrogier, thank you very much for the answer.

I have another question related to ResultSet. The example below works fine.

Statement st(con);
st.Execute(otext("select * from test_fetch"));
Resultset rs = st.GetResultset();

I am writing a wrapper for ResultSet 

class ResultSetWrapper
{
public:
ResultSetWrapper(Statment stm)
{
   m_result = std::make_unique<ResultSet>(stm.GetResultSet());
  //in this context the m_result is valid. If I call m_result->IsNull() is returns false.
}

private:
std::uniqur_ptr<ResultSet> m_result = nullptr;
}

Statment stm(conn);
ResultSetWrapper res (stm)

//Now the SmartHandler pointer inside ResultSet(m_result) is null and I am not able to use the result set.

The wrapper is required in my project, cannot use ResultSet class directly.

Please help.

**Reply by ranganathanrajeshkumar** [reply]
<img width="900" alt="Pic" src="https://github.com/vrogier/ocilib/assets/54272377/674b2cbb-1994-4dec-9d38-8ff11b9cdd9e">

**Reply by vrogier** [reply]
A resulset is a child object of a statement object and lives within the scope of its parent.
A resultet  is a virtual concept. It does not exists on oracle side as data is retrieved directly from the statement itself.
Thus you cannot use a resultset if the statement is freed or reexecuted.

**Reply by ranganathanrajeshkumar** [reply]
Thank @vrogier , Yes RecordSet lifetime depends on Statment class, now its working.

**Comment by ranganathanrajeshkumar** [comment]
Now I am getting linker error when I use Bind() method

Statement st(con);
 st.Prepare("UPDATE NAME SET PHONE=:lpn WHERE FIRSTNAME='ROB' RETURNING ID_NEW INTO :code");
 st.Register<int>(":code");
 st.Bind(":lph", "1234567890", BindInfo::In);
 st.ExecutePrepared();

The lib and include files are configured correctly, But only when I use Bind() method I can see this linker error.

This linker error occurs only when I pass 2nd parameter as ostring in the bind() method.

1>ocilibTest.obj : error LNK2019: unresolved external symbol "public: void __cdecl ocilib::Statement::Bind<char const [11]>(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,char const (&)[11],class ocilib::core::Enum<enum ocilib::BindInfo::BindDirectionValues>)" (??$Bind@$$BY0L@$$CBD@Statement@ocilib@@QEAAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAY0L@$$CBDV?$Enum@W4BindDirectionValues@BindInfo@ocilib@@@core@1@@Z) referenced in function main
1>H:\Project\ocilibTest\x64\Debug\ocilibTest.exe : fatal error LNK1120: 1 unresolved externals

**Reply by ranganathanrajeshkumar** [reply]
ok I the method should be used as below:

 st.Bind< ostring, unsigned int>(":fn", fn, fn.length(), BindInfo::In);
